enterprise/wsproxy/proxy.go

```diff
@@ -0,0 +1,251 @@
+package wsproxy
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"reflect"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/coder/coder/coderd/httpapi"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/buildinfo"
+	"github.com/coder/coder/coderd/httpmw"
+	"github.com/coder/coder/coderd/tracing"
+	"github.com/coder/coder/coderd/workspaceapps"
+	"github.com/coder/coder/coderd/wsconncache"
+	"github.com/coder/coder/codersdk"
+	"github.com/coder/coder/enterprise/wsproxy/wsproxysdk"
+)
+
+type Options struct {
+	Logger slog.Logger
+
+	// PrimaryAccessURL is the URL of the primary coderd instance.
+	// This also serves as the DashboardURL.
+	PrimaryAccessURL *url.URL
+	// AccessURL is the URL of the WorkspaceProxy. This is the url to communicate
+	// with this server.
+	AccessURL *url.URL
+
+	// TODO: @emyrk We use these two fields in many places with this comment.
+	//		Maybe we should make some shared options struct?
+	// AppHostname should be the wildcard hostname to use for workspace
+	// applications INCLUDING the asterisk, (optional) suffix and leading dot.
+	// It will use the same scheme and port number as the access URL.
+	// E.g. "*.apps.coder.com" or "*-apps.coder.com".
+	AppHostname string
+	// AppHostnameRegex contains the regex version of options.AppHostname as
+	// generated by httpapi.CompileHostnamePattern(). It MUST be set if
+	// options.AppHostname is set.
+	AppHostnameRegex *regexp.Regexp
+
+	RealIPConfig *httpmw.RealIPConfig
+	// TODO: @emyrk this key needs to be provided via a file or something?
+	//		Maybe we should curl it from the primary over some secure connection?
+	AppSecurityKey workspaceapps.SecurityKey
+
+	Tracing            trace.TracerProvider
+	PrometheusRegistry *prometheus.Registry
+
+	APIRateLimit     int
+	SecureAuthCookie bool
+	DisablePathApps  bool
+
+	ProxySessionToken string
+}
+
+func (o *Options) Validate() error {
+	var errs optErrors
+
+	errs.Required("Logger", o.Logger)
+	errs.Required("PrimaryAccessURL", o.PrimaryAccessURL)
+	errs.Required("AccessURL", o.AccessURL)
+	errs.Required("RealIPConfig", o.RealIPConfig)
+	errs.Required("PrometheusRegistry", o.PrometheusRegistry)
+	errs.NotEmpty("ProxySessionToken", o.ProxySessionToken)
+	errs.NotEmpty("AppSecurityKey", o.AppSecurityKey)
+
+	if len(errs) > 0 {
+		return errs
+	}
+	return nil
+}
+
+// Server is an external workspace proxy server. This server can communicate
+// directly with a workspace. It requires a primary coderd to establish a said
+// connection.
+type Server struct {
```

Comment:
Should this file be called `server.go` instead? It took me a sec to find this.

Reply:
We have `coderd.API` in a file called `coderd.go` so this matches that convention from the sibling `coderd` package

Reply:
I think it signifies the entrypoint of the package, and so if this is that, then maybe we call this `wsproxy.go` instead.

Reply:
Yep, already done as per your other comment!